feat: add `allowProperties` option to require-atomic-updates (#15238)
eslint/eslint: docs/rules/require-atomic-updates.md

## Rule Details

This rule aims to report assignments to variables or properties in cases where the assignments may be based on outdated values.

### Variables

This rule reports an assignment to a variable when it detects the following execution flow in a generator or async function:

1. The variable is read.
2. A `yield` or `await` pauses the function.

> 3. After the function is resumed, a value is assigned to the variable.

Reviewer: For extra clarity:

```suggestion
3. After the function is resumed, a value is assigned to the variable from step 1.
```

Author: Much better, thanks!

> The assignment in step 3 is reported because this flow indicates that the assignment is based on the possibly outdated value of the variable from step 1, as the variable may have been reassigned elsewhere while the function was paused in step 2.

Reviewer: I think we can simplify this a bit:

```suggestion
The assignment in step 3 is reported because it may be incorrectly calculated because the value of the variable from step 1 may have changed between steps 2 and 3.
```

Reviewer: It may also be helpful to explicitly call out the difference between local and global variables hazards with regards to this rule.

Author: > because it may be incorrectly calculated

I wanted to avoid the word "calculated", as that may be interpreted as if there must be a non-trivial expression that calculates the new value, and that the old value must be directly used in that expression (for example, `result = result + await something`). We had complaints that assignments such as `x = "foo"` should never be reported because there is no calculation, and the old value does not contribute to the new value. This rule reports any `read x -> await -> write x` flow, even if the write is a trivial `x = "foo"`, because the old value may have been used in a condition that determines whether or not `x = "foo"` should be executed.

> may have changed between steps 2 and 3

This could be interpreted as:

```
x; // step 1

await something; // step 2

/*
    some code between step 2 and step 3, `x` is changed here
*/

x = y; // step 3
```

I wanted to emphasize that `x` may have changed outside of this execution flow, while it was awaiting in step 2.

Reviewer: Maybe “resolved” then?

Author: I updated this paragraph now, please take a look.

> Note that the rule does not report the aforementioned assignment in any of the following cases:

Reviewer: ```suggestion
Note that the rule does not report the assignment in step 3 in any of the following cases:
```

* If the variable is read again between steps 2 and 3.

> * If the variable cannot be accessed while the function is paused (e.g., if it's a local variable).

Reviewer: ```suggestion
* If the variable cannot be accessed while the function is paused (for example, if it's a local variable).
```

/* eslint require-atomic-updates: error */

let result;

async function foobar() {
    result = await something + result;
}

async function baz() {
    const tmp = doSomething(await somethingElse);
    result += tmp;
}

async function qux() {
    if (!result) {
        const tmp = await initialize();
        if (!result) {
            result = tmp;
        }
    }
}

async function quux() {
    let localVariable = 0;
    localVariable += await something;
}

function* generator() {
    result = (yield) + result;
}
```

### Properties

This rule reports an assignment to a property through a variable when it detects the following execution flow in a generator or async function:

> 1. The variable is read. This can be reading a property through the variable.

Reviewer: This isn’t clear. Reading a variable from a property? Reading a property and storing it in a variable?

Author: This aims to say that `obj` or `obj.prop` can be step 1 for `obj` variable.

```js
obj; // step 1
await something; // step 2
obj.bar = 1; // step 3, error
```

```js
obj.foo; // step 1
await something; // step 2
obj.bar = 1; // step 3, error
```

Reviewer: Maybe this then?

```suggestion
1. The variable or object property is read.
```
